### ISSUE TITLE
cannot cross compile for arm

### ISSUE THREAD

**Issue by mryvz** [actor-framework/actor-framework#622]
hi,
i ve been triying to compile caf for arm embedded linux on ubuntu. make command  fails with error /root/actor-framework-0.15.5/libcaf_core/caf/scheduled_actor.hpp:115:59: error: 'exception_ptr' is not a member of 'std'
   using exception_handler = std::function<error (pointer, std::exception_ptr&)>;
...
anyone has a solution ?

gcc info

```
root@659d2e0d753c:~/actor-framework-0.15.5# arm-linux-gnueabi-g++-5 -v
Using built-in specs.
COLLECT_GCC=arm-linux-gnueabi-g++-5
COLLECT_LTO_WRAPPER=/usr/lib/gcc-cross/arm-linux-gnueabi/5/lto-wrapper
Target: arm-linux-gnueabi
Configured with: ../src/configure -v --with-pkgversion='Ubuntu/Linaro 5.4.0-6ubuntu1~16.04.4' --with-bugurl=file:///usr/share/doc/gcc-5/README.Bugs --enable-languages=c,ada
,c++,java,go,d,fortran,objc,obj-c++ --prefix=/usr --program-suffix=-5 --enable-shared --enable-linker-build-id --libexecdir=/usr/lib --without-included-gettext --enable-thr
eads=posix --libdir=/usr/lib --enable-nls --with-sysroot=/ --enable-clocale=gnu --enable-libstdcxx-debug --enable-libstdcxx-time=yes --with-default-libstdcxx-abi=new --enab
le-gnu-unique-object --disable-libitm --disable-libquadmath --enable-plugin --with-system-zlib --disable-browser-plugin --enable-java-awt=gtk --enable-gtk-cairo --with-java
-home=/usr/lib/jvm/java-1.5.0-gcj-5-armel-cross/jre --enable-java-home --with-jvm-root-dir=/usr/lib/jvm/java-1.5.0-gcj-5-armel-cross --with-jvm-jar-dir=/usr/lib/jvm-exports
/java-1.5.0-gcj-5-armel-cross --with-arch-directory=arm --with-ecj-jar=/usr/share/java/eclipse-ecj.jar --disable-libgcj --enable-objc-gc --enable-multiarch --enable-multili
b --disable-sjlj-exceptions --with-arch=armv5t --with-float=soft --disable-werror --enable-multilib --enable-checking=release --build=x86_64-linux-gnu --host=x86_64-linux-g
nu --target=arm-linux-gnueabi --program-prefix=arm-linux-gnueabi- --includedir=/usr/arm-linux-gnueabi/include
Thread model: posix
gcc version 5.4.0 20160609 (Ubuntu/Linaro 5.4.0-6ubuntu1~16.04.4)
```

./configure output
...

```

-- Looking for pthread.h
-- Looking for pthread.h - found
-- Looking for pthread_create
-- Looking for pthread_create - not found
-- Looking for pthread_create in pthreads
-- Looking for pthread_create in pthreads - not found
-- Looking for pthread_create in pthread
-- Looking for pthread_create in pthread - found
-- Found Threads: TRUE
-- Could NOT find Protobuf (missing:  PROTOBUF_LIBRARY PROTOBUF_INCLUDE_DIR)
-- Found unsuitable Qt version "" from NOTFOUND
-- Could NOT find CURL (missing:  CURL_LIBRARY CURL_INCLUDE_DIR)
-- Found 61 test suites
-- Could NOT find Doxygen (missing:  DOXYGEN_EXECUTABLE)
--
====================|  Build Summary  |====================

Libcaf version:        0.15.5

Build type:            RelWithDebInfo
Build static:          no
Build static only:     no
Build static runtime:
Runtime checks:        no
Log level:             none
With mem. mgmt.:       yes
With exceptions:       yes

Build I/O module:      yes
Build tools:           yes
Build examples:        yes
Build unit tests:      yes
Build benchmarks:      no
Build OpenCL:          no
Build OpenSSL:         no
Build Python:          yes

CXX:                   /usr/bin/arm-linux-gnueabi-g++-5
CXXFLAGS:              -mtune=arm9  -std=c++11 -Wextra -Wall -pedantic -Wno-missing-field-initializers -pthread -fPIC -O2 -g -DNDEBUG  -O2 -g -DNDEBUG
LIBRARIES:

Source directory:      /root/actor-framework-0.15.5
Build directory:       /root/actor-framework-0.15.5/build
Executable path:       /root/actor-framework-0.15.5/build/bin
Library path:          /root/actor-framework-0.15.5/build/lib
Install prefix:        /usr/local
Generator:             Unix Makefiles
```

make output 
```
  1%] Building CXX object libcaf_core/CMakeFiles/libcaf_core_shared.dir/src/abstract_coordinator.cpp.o
Scanning dependencies of target libcaf_core
make[3]: Leaving directory '/root/actor-framework-0.15.5/build'
[  0%] Built target libcaf_core
make[3]: Entering directory '/root/actor-framework-0.15.5/build'
Scanning dependencies of target libcaf_core_shared
make[3]: Leaving directory '/root/actor-framework-0.15.5/build'
make[3]: Entering directory '/root/actor-framework-0.15.5/build'
[  1%] Building CXX object libcaf_core/CMakeFiles/libcaf_core_shared.dir/src/abstract_actor.cpp.o
[  1%] Building CXX object libcaf_core/CMakeFiles/libcaf_core_shared.dir/src/abstract_channel.cpp.o
[  2%] Building CXX object libcaf_core/CMakeFiles/libcaf_core_shared.dir/src/abstract_composable_behavior.cpp.o
[  2%] Building CXX object libcaf_core/CMakeFiles/libcaf_core_shared.dir/src/abstract_coordinator.cpp.o
In file included from /root/actor-framework-0.15.5/libcaf_core/caf/typed_actor_view.hpp:23:0,
                 from /root/actor-framework-0.15.5/libcaf_core/caf/typed_actor_pointer.hpp:23,
                 from /root/actor-framework-0.15.5/libcaf_core/caf/actor_ostream.hpp:25,
                 from /root/actor-framework-0.15.5/libcaf_core/src/abstract_coordinator.cpp:37:
/root/actor-framework-0.15.5/libcaf_core/caf/scheduled_actor.hpp:115:59: error: 'exception_ptr' is not a member of 'std'
   using exception_handler = std::function<error (pointer, std::exception_ptr&)>;
                                                           ^
/root/actor-framework-0.15.5/libcaf_core/caf/scheduled_actor.hpp:115:59: error: 'exception_ptr' is not a member of 'std'
/root/actor-framework-0.15.5/libcaf_core/caf/scheduled_actor.hpp:115:79: error: template argument 1 is invalid
   using exception_handler = std::function<error (pointer, std::exception_ptr&)>;
                                                                               ^
/root/actor-framework-0.15.5/libcaf_core/caf/scheduled_actor.hpp:157:60: error: 'std::exception_ptr' has not been declared
   static error default_exception_handler(pointer ptr, std::exception_ptr& x);
                                                            ^
/root/actor-framework-0.15.5/libcaf_core/caf/scheduled_actor.hpp:286:37: error: 'exception_handler' has not been declared
   inline void set_exception_handler(exception_handler fun) {
                                     ^
/root/actor-framework-0.15.5/libcaf_core/caf/scheduled_actor.hpp:299:28: error: 'exception_ptr' is not a member of 'std'
       std::function<error (std::exception_ptr&)>
                            ^
/root/actor-framework-0.15.5/libcaf_core/caf/scheduled_actor.hpp:299:28: error: 'exception_ptr' is not a member of 'std'
/root/actor-framework-0.15.5/libcaf_core/caf/scheduled_actor.hpp:299:48: error: template argument 1 is invalid
       std::function<error (std::exception_ptr&)>
                                                ^
/root/actor-framework-0.15.5/libcaf_core/caf/scheduled_actor.hpp:300:5: error: template argument 2 is invalid
     >::value
     ^
/root/actor-framework-0.15.5/libcaf_core/caf/scheduled_actor.hpp:301:3: error: template argument 1 is invalid
   >::type
   ^
/root/actor-framework-0.15.5/libcaf_core/caf/scheduled_actor.hpp:877:3: error: 'exception_handler' does not name a type
   exception_handler exception_handler_;
   ^
/root/actor-framework-0.15.5/libcaf_core/caf/scheduled_actor.hpp: In member function 'void caf::scheduled_actor::set_exception_handler(int)':
/root/actor-framework-0.15.5/libcaf_core/caf/scheduled_actor.hpp:288:7: error: 'exception_handler_' was not declared in this scope
       exception_handler_ = std::move(fun);
       ^
/root/actor-framework-0.15.5/libcaf_core/caf/scheduled_actor.hpp:290:7: error: 'exception_handler_' was not declared in this scope
       exception_handler_ = default_exception_handler;
       ^
/root/actor-framework-0.15.5/libcaf_core/caf/scheduled_actor.hpp: In member function 'int caf::scheduled_actor::set_exception_handler(F)':
/root/actor-framework-0.15.5/libcaf_core/caf/scheduled_actor.hpp:303:54: error: 'std::exception_ptr' has not been declared
     set_exception_handler([f](scheduled_actor*, std::exception_ptr& x) {
                                                      ^
/root/actor-framework-0.15.5/libcaf_core/caf/scheduled_actor.hpp:306:3: warning: no return statement in function returning non-void [-Wreturn-type]
   }
   ^
libcaf_core/CMakeFiles/libcaf_core_shared.dir/build.make:134: recipe for target 'libcaf_core/CMakeFiles/libcaf_core_shared.dir/src/abstract_coordinator.cpp.o' failed
make[3]: *** [libcaf_core/CMakeFiles/libcaf_core_shared.dir/src/abstract_coordinator.cpp.o] Error 1
make[3]: Leaving directory '/root/actor-framework-0.15.5/build'
CMakeFiles/Makefile2:285: recipe for target 'libcaf_core/CMakeFiles/libcaf_core_shared.dir/all' failed
make[2]: *** [libcaf_core/CMakeFiles/libcaf_core_shared.dir/all] Error 2
make[2]: Leaving directory '/root/actor-framework-0.15.5/build'
Makefile:138: recipe for target 'all' failed
make[1]: *** [all] Error 2
make[1]: Leaving directory '/root/actor-framework-0.15.5/build'
Makefile:4: recipe for target 'all' failed
make: *** [all] Error 2
```

**Comment by Hamdor** [comment]
Hello,

you can try to disable exceptions. This can be done with `./configure --no-exceptions`.

Best regards,
Marian

Am 11.11.2017 um 10:13 schrieb mryvz <notifications@github.com<mailto:notifications@github.com>>:

hi,
i ve been triying to compile caf for arm embedded linux on ubuntu. make command fails with error /root/actor-framework-0.15.5/libcaf_core/caf/scheduled_actor.hpp:115:59: error: 'exception_ptr' is not a member of 'std'
using exception_handler = std::function<error (pointer, std::exception_ptr&)>;
...
anyone has a solution ?

gcc info

root@659d2e0d753c:~/actor-framework-0.15.5# arm-linux-gnueabi-g++-5 -v
Using built-in specs.
COLLECT_GCC=arm-linux-gnueabi-g++-5
COLLECT_LTO_WRAPPER=/usr/lib/gcc-cross/arm-linux-gnueabi/5/lto-wrapper
Target: arm-linux-gnueabi
Configured with: ../src/configure -v --with-pkgversion='Ubuntu/Linaro 5.4.0-6ubuntu1~16.04.4' --with-bugurl=file:///usr/share/doc/gcc-5/README.Bugs --enable-languages=c,ada
,c++,java,go,d,fortran,objc,obj-c++ --prefix=/usr --program-suffix=-5 --enable-shared --enable-linker-build-id --libexecdir=/usr/lib --without-included-gettext --enable-thr
eads=posix --libdir=/usr/lib --enable-nls --with-sysroot=/ --enable-clocale=gnu --enable-libstdcxx-debug --enable-libstdcxx-time=yes --with-default-libstdcxx-abi=new --enab
le-gnu-unique-object --disable-libitm --disable-libquadmath --enable-plugin --with-system-zlib --disable-browser-plugin --enable-java-awt=gtk --enable-gtk-cairo --with-java
-home=/usr/lib/jvm/java-1.5.0-gcj-5-armel-cross/jre --enable-java-home --with-jvm-root-dir=/usr/lib/jvm/java-1.5.0-gcj-5-armel-cross --with-jvm-jar-dir=/usr/lib/jvm-exports
/java-1.5.0-gcj-5-armel-cross --with-arch-directory=arm --with-ecj-jar=/usr/share/java/eclipse-ecj.jar --disable-libgcj --enable-objc-gc --enable-multiarch --enable-multili
b --disable-sjlj-exceptions --with-arch=armv5t --with-float=soft --disable-werror --enable-multilib --enable-checking=release --build=x86_64-linux-gnu --host=x86_64-linux-g
nu --target=arm-linux-gnueabi --program-prefix=arm-linux-gnueabi- --includedir=/usr/arm-linux-gnueabi/include
Thread model: posix
gcc version 5.4.0 20160609 (Ubuntu/Linaro 5.4.0-6ubuntu1~16.04.4)

./configure output
...


-- Looking for pthread.h
-- Looking for pthread.h - found
-- Looking for pthread_create
-- Looking for pthread_create - not found
-- Looking for pthread_create in pthreads
-- Looking for pthread_create in pthreads - not found
-- Looking for pthread_create in pthread
-- Looking for pthread_create in pthread - found
-- Found Threads: TRUE
-- Could NOT find Protobuf (missing:  PROTOBUF_LIBRARY PROTOBUF_INCLUDE_DIR)
-- Found unsuitable Qt version "" from NOTFOUND
-- Could NOT find CURL (missing:  CURL_LIBRARY CURL_INCLUDE_DIR)
-- Found 61 test suites
-- Could NOT find Doxygen (missing:  DOXYGEN_EXECUTABLE)
--
====================|  Build Summary  |====================

Libcaf version:        0.15.5

Build type:            RelWithDebInfo
Build static:          no
Build static only:     no
Build static runtime:
Runtime checks:        no
Log level:             none
With mem. mgmt.:       yes
With exceptions:       yes

Build I/O module:      yes
Build tools:           yes
Build examples:        yes
Build unit tests:      yes
Build benchmarks:      no
Build OpenCL:          no
Build OpenSSL:         no
Build Python:          yes

CXX:                   /usr/bin/arm-linux-gnueabi-g++-5
CXXFLAGS:              -mtune=arm9  -std=c++11 -Wextra -Wall -pedantic -Wno-missing-field-initializers -pthread -fPIC -O2 -g -DNDEBUG  -O2 -g -DNDEBUG
LIBRARIES:

Source directory:      /root/actor-framework-0.15.5
Build directory:       /root/actor-framework-0.15.5/build
Executable path:       /root/actor-framework-0.15.5/build/bin
Library path:          /root/actor-framework-0.15.5/build/lib
Install prefix:        /usr/local
Generator:             Unix Makefiles

make output

  1%] Building CXX object libcaf_core/CMakeFiles/libcaf_core_shared.dir/src/abstract_coordinator.cpp.o
Scanning dependencies of target libcaf_core
make[3]: Leaving directory '/root/actor-framework-0.15.5/build'
[  0%] Built target libcaf_core
make[3]: Entering directory '/root/actor-framework-0.15.5/build'
Scanning dependencies of target libcaf_core_shared
make[3]: Leaving directory '/root/actor-framework-0.15.5/build'
make[3]: Entering directory '/root/actor-framework-0.15.5/build'
[  1%] Building CXX object libcaf_core/CMakeFiles/libcaf_core_shared.dir/src/abstract_actor.cpp.o
[  1%] Building CXX object libcaf_core/CMakeFiles/libcaf_core_shared.dir/src/abstract_channel.cpp.o
[  2%] Building CXX object libcaf_core/CMakeFiles/libcaf_core_shared.dir/src/abstract_composable_behavior.cpp.o
[  2%] Building CXX object libcaf_core/CMakeFiles/libcaf_core_shared.dir/src/abstract_coordinator.cpp.o
In file included from /root/actor-framework-0.15.5/libcaf_core/caf/typed_actor_view.hpp:23:0,
                 from /root/actor-framework-0.15.5/libcaf_core/caf/typed_actor_pointer.hpp:23,
                 from /root/actor-framework-0.15.5/libcaf_core/caf/actor_ostream.hpp:25,
                 from /root/actor-framework-0.15.5/libcaf_core/src/abstract_coordinator.cpp:37:
/root/actor-framework-0.15.5/libcaf_core/caf/scheduled_actor.hpp:115:59: error: 'exception_ptr' is not a member of 'std'
   using exception_handler = std::function<error (pointer, std::exception_ptr&)>;
                                                           ^
/root/actor-framework-0.15.5/libcaf_core/caf/scheduled_actor.hpp:115:59: error: 'exception_ptr' is not a member of 'std'
/root/actor-framework-0.15.5/libcaf_core/caf/scheduled_actor.hpp:115:79: error: template argument 1 is invalid
   using exception_handler = std::function<error (pointer, std::exception_ptr&)>;
                                                                               ^
/root/actor-framework-0.15.5/libcaf_core/caf/scheduled_actor.hpp:157:60: error: 'std::exception_ptr' has not been declared
   static error default_exception_handler(pointer ptr, std::exception_ptr& x);
                                                            ^
/root/actor-framework-0.15.5/libcaf_core/caf/scheduled_actor.hpp:286:37: error: 'exception_handler' has not been declared
   inline void set_exception_handler(exception_handler fun) {
                                     ^
/root/actor-framework-0.15.5/libcaf_core/caf/scheduled_actor.hpp:299:28: error: 'exception_ptr' is not a member of 'std'
       std::function<error (std::exception_ptr&)>
                            ^
/root/actor-framework-0.15.5/libcaf_core/caf/scheduled_actor.hpp:299:28: error: 'exception_ptr' is not a member of 'std'
/root/actor-framework-0.15.5/libcaf_core/caf/scheduled_actor.hpp:299:48: error: template argument 1 is invalid
       std::function<error (std::exception_ptr&)>
                                                ^
/root/actor-framework-0.15.5/libcaf_core/caf/scheduled_actor.hpp:300:5: error: template argument 2 is invalid
     >::value
     ^
/root/actor-framework-0.15.5/libcaf_core/caf/scheduled_actor.hpp:301:3: error: template argument 1 is invalid
   >::type
   ^
/root/actor-framework-0.15.5/libcaf_core/caf/scheduled_actor.hpp:877:3: error: 'exception_handler' does not name a type
   exception_handler exception_handler_;
   ^
/root/actor-framework-0.15.5/libcaf_core/caf/scheduled_actor.hpp: In member function 'void caf::scheduled_actor::set_exception_handler(int)':
/root/actor-framework-0.15.5/libcaf_core/caf/scheduled_actor.hpp:288:7: error: 'exception_handler_' was not declared in this scope
       exception_handler_ = std::move(fun);
       ^
/root/actor-framework-0.15.5/libcaf_core/caf/scheduled_actor.hpp:290:7: error: 'exception_handler_' was not declared in this scope
       exception_handler_ = default_exception_handler;
       ^
/root/actor-framework-0.15.5/libcaf_core/caf/scheduled_actor.hpp: In member function 'int caf::scheduled_actor::set_exception_handler(F)':
/root/actor-framework-0.15.5/libcaf_core/caf/scheduled_actor.hpp:303:54: error: 'std::exception_ptr' has not been declared
     set_exception_handler([f](scheduled_actor*, std::exception_ptr& x) {
                                                      ^
/root/actor-framework-0.15.5/libcaf_core/caf/scheduled_actor.hpp:306:3: warning: no return statement in function returning non-void [-Wreturn-type]
   }
   ^
libcaf_core/CMakeFiles/libcaf_core_shared.dir/build.make:134: recipe for target 'libcaf_core/CMakeFiles/libcaf_core_shared.dir/src/abstract_coordinator.cpp.o' failed
make[3]: *** [libcaf_core/CMakeFiles/libcaf_core_shared.dir/src/abstract_coordinator.cpp.o] Error 1
make[3]: Leaving directory '/root/actor-framework-0.15.5/build'
CMakeFiles/Makefile2:285: recipe for target 'libcaf_core/CMakeFiles/libcaf_core_shared.dir/all' failed
make[2]: *** [libcaf_core/CMakeFiles/libcaf_core_shared.dir/all] Error 2
make[2]: Leaving directory '/root/actor-framework-0.15.5/build'
Makefile:138: recipe for target 'all' failed
make[1]: *** [all] Error 2
make[1]: Leaving directory '/root/actor-framework-0.15.5/build'
Makefile:4: recipe for target 'all' failed
make: *** [all] Error 2

**Comment by mryvz** [comment]
hello again,
> you can try to disable exceptions. This can be done with `./configure --no-exceptions`.

--no-exceptions worked , saved my day.

 thank you very much
Best Regards
Ömür
